### PR TITLE
Add storage warning banner to the top of overview page

### PR DESCRIPTION
**Pull request by p-jackson**
Fixes DOTDASH-138

## Proposed Changes

* Render `<StorageWarningBanner>` at top of overview page
* Ensure all necessary data is loaded before rendering the overview
  * Prevents layout shift
* "Low storage" warnings can be dismissed
* Move logic that decides what counts as "low" storage into a shared function
* Tidy up: remove the `'some-string'` placeholder from the `UserPreferences` type. It was only there to check the types were working nicely.
* Accessible label for `<Notice>` components dismiss button
* Unit tests

**Site has used over 80%**

CleanShot 2025-08-04 at 16 31 22@2x

**Site has used over 100%**

CleanShot 2025-08-04 at 16 32 20@2x

## Why are these changes being made?

Pretty sure this design supersedes this figma QOBjujZah0UlGDDdLKZhzi-fi-7256_103340 which shows the link in the plan card. I can't find the P2 now, but we discussed the warning for low-storage needing to be more prominent because when it gets too low backups can start failing. CC @matt-west 

## Testing Instructions

I found the easiest way to test this was simply to hardcode the new `getStorageAlertLevel()` function to return the value I wanted.

The easiest way to clear any old preferences is to use the dev menu in v1

CleanShot 2025-08-04 at 16 30 06@2x

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you tested accessibility for your changes? Ensure the feature remains usable with various user agents (e.g., browsers), interfaces (e.g., keyboard navigation), and assistive technologies (e.g., screen readers) (PCYsg-S3g-p2).
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

File: client/dashboard/sites/overview/storage-warning-banner.tsx [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+			actions={
+				<UpsellCTAButton variant="primary" tracksId={ tracksId } href={ `/add-ons/${ site.slug }` }>
+					{ __( 'Add more storage' ) }
+				</UpsellCTAButton>
+			}
+		>
+			{ sprintf(
+				// translators: "used" and "available" amounts data including a unit, e.g. "2.03 MB" or "1.5 GB".
+				__(
+					'%(used)s of your %(available)s storage limit has been used. Upgrade to continue storing media, plugins, themes, and backups.'
```
Comment (arthur791004) [inline client/dashboard/sites/overview/storage-warning-banner.tsx:57]: Would it be better to display the percentage of the usage cc @matt-west  🤔

File: client/dashboard/sites/overview/test/storage-warning-banner.tsx [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+		max_storage_bytes,
+		max_storage_bytes_from_addons: 0,
+	};
+
+	const preferences = {
+		[ `hosting-dashboard-overview-storage-notice-dismissed-${ siteId }` ]:
+			typeof isDismissed === 'string' ? isDismissed : isDismissed && 'today',
+	};
+
+	const scope = nock( 'https://public-api.wordpress.com' )
```
Comment (arthur791004) [inline client/dashboard/sites/overview/test/storage-warning-banner.tsx:34]: [nit] Maybe we can try msw as it looks better and has better integration with the storybook 😄

File: client/dashboard/sites/overview/test/storage-warning-banner.tsx [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+test( 'no banner when storage is not low', async () => {
+	mockTestEndpoints( {
+		storage_used_bytes: 100,
+		max_storage_bytes: 5000,
+	} );
+
+	renderTestBanner();
+
+	await waitFor( () => expect( screen.queryByText( 'Loading...' ) ).not.toBeInTheDocument() );
```
Comment (arthur791004) [inline client/dashboard/sites/overview/test/storage-warning-banner.tsx]: Do we need to test it 🤔

File: client/dashboard/test-utils.tsx [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+		defaultOptions: {
+			queries: { retry: false },
+		},
+	} );
+	const router = createRouter( {
+		routeTree: createRootRoute( { component: () => ui } ),
+	} );
+	return testingLibraryRender(
+		<QueryClientProvider client={ queryClient }>
+			<Suspense fallback={ <div>Loading...</div> }>
```
Comment (arthur791004) [inline client/dashboard/test-utils.tsx]: Doesn't the `RouterProvider` render the Suspense by default 🤔

File: client/dashboard/utils/site-storage.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+const ALERT_FRACTION = 0.8;
+
+export function getStorageAlertLevel( {
+	storage_used_bytes,
+	max_storage_bytes,
+}: SiteMediaStorage ): 'none' | 'low' | 'exceeded' {
+	const storageFraction = storage_used_bytes / max_storage_bytes;
+
+	if ( storageFraction > 1 ) {
```
Comment (arthur791004) [inline client/dashboard/utils/site-storage.ts:11]: Do we need to include `=`?

File: client/dashboard/sites/overview/storage-warning-banner.tsx [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+	const { data: mediaStorage } = useSuspenseQuery( siteMediaStorageQuery( site.ID ) );
+	const { data: isDismissedPersisted } = useSuspenseQuery(
+		userPreferenceQuery( `hosting-dashboard-overview-storage-notice-dismissed-${ site.ID }` )
+	);
+	const { mutate: updateDismissed, isPending: isDismissing } = useMutation(
+		userPreferenceMutation( `hosting-dashboard-overview-storage-notice-dismissed-${ site.ID }` )
+	);
+
+	// Optimistically hide the banner assuming the preference will get saved.
+	const isDismissed = isDismissedPersisted || isDismissing;
```
Comment (arthur791004) [inline client/dashboard/sites/overview/storage-warning-banner.tsx:21]: Is it possible to show the dismissed banner again if the storage on the new plan is about to run out? 🤔

File: client/dashboard/sites/overview/storage-warning-banner.tsx [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+import { getStorageAlertLevel } from '../../utils/site-storage';
+import type { Site } from '../../data/types';
+
+export function StorageWarningBanner( { site }: { site: Site } ) {
+	const { data: mediaStorage } = useSuspenseQuery( siteMediaStorageQuery( site.ID ) );
+	const { data: isDismissedPersisted } = useSuspenseQuery(
+		userPreferenceQuery( `hosting-dashboard-overview-storage-notice-dismissed-${ site.ID }` )
+	);
+	const { mutate: updateDismissed, isPending: isDismissing } = useMutation(
+		userPreferenceMutation( `hosting-dashboard-overview-storage-notice-dismissed-${ site.ID }` )
```
Comment (arthur791004) [inline client/dashboard/sites/overview/storage-warning-banner.tsx:17]: We’ve had similar implementation previously, but I’m wondering what would happen if there are many site-related preferences. It seems like this could potentially slow down the API 🤔

File: client/dashboard/utils/site-storage.ts [hunk 8]
```diff
@@ -0,0 +1,17 @@
+import type { SiteMediaStorage } from '../data/types';
+
+const ALERT_FRACTION = 0.8;
+
+export function getStorageAlertLevel( {
+	storage_used_bytes,
+	max_storage_bytes,
+}: SiteMediaStorage ): 'none' | 'low' | 'exceeded' {
```
Comment (arthur791004) [inline client/dashboard/utils/site-storage.ts]: Would it be better to call it `warning` instead of `low`?

File: client/dashboard/sites/overview/storage-warning-banner.tsx (hunk 1 above)
Comment (matt-west) [reply client/dashboard/sites/overview/storage-warning-banner.tsx:57]: The percentage would work too. As long as we still include what their limit is.

File: client/dashboard/app/router.tsx [hunk 9]
```diff
@@ -129,6 +130,8 @@ const siteOverviewRoute = createRoute( {
 	path: '/',
 	loader: async ( { params: { siteSlug }, preload } ) => {
 		const site = await queryClient.ensureQueryData( siteBySlugQuery( siteSlug ) );
+		// Ensure storage specifically is loaded because the warning notice can cause a layout shift
+		await queryClient.ensureQueryData( siteMediaStorageQuery( site.ID ) );
```
Comment (fushar) [inline client/dashboard/app/router.tsx]: I'm torn and not totally sure about this... currently, sometimes clicking on a site row is already slow because the site endpoint slow sometimes. I think layout shift for some minority sites is a reasonable tradeoff so that it won't hurt majority cases 🫠 (also it will only happen when the cache is empty)

File: client/dashboard/sites/overview/storage-warning-banner.tsx (hunk 6 above)
Comment (fushar) [reply client/dashboard/sites/overview/storage-warning-banner.tsx:21]: I think one solution is to show a "Add more storage" link/button in the Plan card is case the site is low on storage. So that the user can take action even though they might have dismissed the banner. cc: @matt-west 

Image

File: client/dashboard/utils/site-storage.ts (hunk 8 above)
Comment (fushar) [reply client/dashboard/utils/site-storage.ts]: Yeah, in any case, `low` and `exceeded` seem contradictory. One is about remaining space and one is about used space.

File: client/dashboard/sites/overview/test/storage-warning-banner.tsx (hunk 3 above)
Comment (fushar) [reply client/dashboard/sites/overview/test/storage-warning-banner.tsx]: I think it's to wait until the loading state has finished.

File: client/dashboard/test-utils.tsx (hunk 4 above)
Comment (fushar) [reply client/dashboard/test-utils.tsx]: I think it's so that we can distinguish loading states in the tests easily 😄

File: client/dashboard/sites/overview/test/storage-warning-banner.tsx (hunk 2 above)
Comment (p-jackson) [reply client/dashboard/sites/overview/test/storage-warning-banner.tsx:34]: Yeah I like the sound of msw, I haven't used it before though. I used nock because it's already used in the Calypso codebase.
I see msw in the `yarn.lock`, so maybe we're already using it too?

File: client/dashboard/sites/overview/test/storage-warning-banner.tsx (hunk 3 above)
Comment (p-jackson) [reply client/dashboard/sites/overview/test/storage-warning-banner.tsx]: Yep. We can rely on `useSuspenseQuery()` in the real app since we have the router. But in test code we have to make sure the component has finished loading before we assert whether something is _not_ there.

File: client/dashboard/test-utils.tsx (hunk 4 above)
Comment (p-jackson) [reply client/dashboard/test-utils.tsx]: Yeah, I want something concrete to test for. Actually it might be better to use a `data-testid` on this element since we know we're testing for such a specific thing.

Afaik there isn't another way to test for "the component is no longer suspended"

File: client/dashboard/utils/site-storage.ts (hunk 5 above)
Comment (p-jackson) [reply client/dashboard/utils/site-storage.ts:11]: Well ... I figure if you're at 100% of your usage, your not technically over your plan right? If you've been allowed 1G and you somehow use precisely 1G then you haven't actually violated the plan right?

File: client/dashboard/sites/overview/storage-warning-banner.tsx (hunk 6 above)
Comment (p-jackson) [reply client/dashboard/sites/overview/storage-warning-banner.tsx:21]: > Is it possible to show the dismissed banner again if the storage on the new plan is about to run out? 🤔

Yeah that's why I stored the dismiss date. I thought in the future we might want to do something like bring it back to re-warn the user.
Maybe we should be persisting more than the date. Maybe we should persist the date _and_ the level that they dismissed the notice at. In the future we could use that to come up with some fancy logic for bringing the banner back.

Comment (p-jackson) [reply client/dashboard/sites/overview/storage-warning-banner.tsx:21]: > So that the user can take action even though they might have dismissed the banner

Oh that's a good point! The figma has that link, but I figured this banner superseded the link. I should add the link too, that way it gives the user a low key way to purchase the addon.

File: client/dashboard/sites/overview/storage-warning-banner.tsx (hunk 7 above)
Comment (p-jackson) [reply client/dashboard/sites/overview/storage-warning-banner.tsx:17]: I think we either keep them all on different keys, or we keep them all on a single key with a `Record<number, boolean>` type. The second is more data compact probably. But I think also more easy to mess up. For instance up here https://github.com/Automattic/wp-calypso/pull/105041#discussion_r2252787475 I was just thinking of messing with the data format. It gets messy when that one key is going to have data with old and new formats.

Storing everything as its own key feels like a KISS approach that's worked for us so far 🤞

File: client/dashboard/utils/site-storage.ts (hunk 8 above)
Comment (p-jackson) [reply client/dashboard/utils/site-storage.ts]: I see yeah 😅 what does `none` mean? "no" storage remaining? But here it means no warning.

I'll switching to `'none' | 'warning' | 'exceeded'`

File: client/dashboard/app/router.tsx (hunk 9 above)
Comment (p-jackson) [reply client/dashboard/app/router.tsx]: I hate layout shift with a burning passion 😅 
If we don't feel we can afford to pre-load the remaining storage then I would rather we switch to a new design which doesn't cause the layout to shift.

File: client/dashboard/sites/overview/storage-warning-banner.tsx (hunk 6 above)
Comment (fushar) [reply client/dashboard/sites/overview/storage-warning-banner.tsx:21]: I think we're supposed to implement both 😄 

image

File: client/dashboard/app/router.tsx (hunk 9 above)
Comment (fushar) [reply client/dashboard/app/router.tsx]: Okay, we can disagree 😄 

Maybe we can optimize the code a bit by starting all the preload Promises right away without await-ing the media storage endpoint? (might require a bit of moving around code)

File: client/dashboard/utils/site-storage.ts (hunk 5 above)
Comment (arthur791004) [reply client/dashboard/utils/site-storage.ts:11]: I guess we can ignore it as it seems to be an edge case 😄

File: client/dashboard/sites/overview/test/storage-warning-banner.tsx (hunk 2 above)
Comment (arthur791004) [reply client/dashboard/sites/overview/test/storage-warning-banner.tsx:34]: I saw launchpad and plan-grid-next use it for the storybook.

File: client/dashboard/sites/overview/test/storage-warning-banner.tsx (hunk 3 above)
Comment (arthur791004) [reply client/dashboard/sites/overview/test/storage-warning-banner.tsx]: Yeah, but it feels a little off that we need it for test cases...

Comment (p-jackson) [reply client/dashboard/sites/overview/test/storage-warning-banner.tsx]: Maybe I should embed the waiting inside the `renderTestBanner()` function. That way we're saying that by the time "render" function returns we're ready to make assertions.

File: client/dashboard/sites/overview/test/storage-warning-banner.tsx (hunk 2 above)
Comment (p-jackson) [reply client/dashboard/sites/overview/test/storage-warning-banner.tsx:34]: I'll take a look at msw in a follow up PR. I've been thinking of doing some work on test coverage, so that seems like a good time to try it out.

File: client/dashboard/test-utils.tsx (hunk 4 above)
Comment (p-jackson) [reply client/dashboard/test-utils.tsx]: > Doesn't the `RouterProvider` render the Suspense by default 🤔

My test router implementation was a bit busted. I've rearranged where the `<Suspense>` is and its working much better now. Thanks for flagging.

File: client/dashboard/sites/overview/storage-warning-banner.tsx (hunk 6 above)
Comment (p-jackson) [reply client/dashboard/sites/overview/storage-warning-banner.tsx:21]: Yes, upon reflection this does make sense 😄

File: client/dashboard/sites/overview/test/storage-warning-banner.tsx (hunk 3 above)
Comment (arthur791004) [reply client/dashboard/sites/overview/test/storage-warning-banner.tsx]: I tested it, and it works without it. Is there something I'm missing?

File: client/dashboard/sites/overview/storage-warning-banner.tsx (hunk 6 above)
Comment (p-jackson) [reply client/dashboard/sites/overview/storage-warning-banner.tsx:21]: Following this figma QOBjujZah0UlGDDdLKZhzi-fi-7256_103340

CleanShot 2025-08-05 at 20 06 32@2x

I'm using an external link as that seems to be the pattern we're folling.
Not that when the plan card expands, the horizontal separator doesn't line up so nicely anymore. But as discussed here—DOTDASH-114#comment-37952b6c—when the plan card height expands, all the other cards are supposed to grow too (as opposed to the plans card hanging lower)

File: client/dashboard/sites/overview/test/storage-warning-banner.tsx (hunk 3 above)
Comment (p-jackson) [reply client/dashboard/sites/overview/test/storage-warning-banner.tsx]: So if you first comment out the code waiting for `data-testid="loading"`

You see that it passes, but I believe this is just because nock is too darn fast and `useSuspenseQuery()` must be returning too quick.

If you add an artificial delay in nock like this:

```ts
	const scope = nock( 'https://public-api.wordpress.com' )
		.get( `/rest/v1.1/sites/${ siteId }/media-storage` )
		.delay( 500 )
		.query( true )
		.reply( 200, mediaStorage )
		.get( '/rest/v1.1/me/preferences' )
		.delay( 500 )
		.query( true )
		.reply( 200, { calypso_preferences: preferences } )
		.post( '/rest/v1.1/me/preferences' )
		.delay( 500 )
		.reply( 200, ( _uri, body ) => body );
```

Then you should see that the test now _does_ fail as expected. It's trying to assert the "warning" behaviour, but it can't because it is loading.

Now you can make the tests pass again by uncommenting the `waitFor` the loading element. Since nock is making everything run much slower you might need to tell `waitFor` to wait longer. e.g.

```ts
	await waitFor( () => expect( screen.queryByTestId( 'loading' ) ).not.toBeInTheDocument(), {
		timeout: 5000,
	} );
```

So I believe that semantically the tests are more correct if we wait for the component to un-suspend. But on a practical level it isn't necessary since the network mocking is fast enough. But IMO it just seems more resilient to not write tests that depend on the quick behaviour.

Comment (arthur791004) [reply client/dashboard/sites/overview/test/storage-warning-banner.tsx]: I don't think we need to consider the network behavior here, since that's not the focus of these tests. What we're really trying to verify is whether the component renders correctly based on the given data. So, I believe we can safely ignore the slow response in this case 🤔

In any case, it's not a blocker for this PR 😄

Comment (fushar) [reply client/dashboard/sites/overview/test/storage-warning-banner.tsx]: I think it's fine. The test is `no banner when storage is not low`, so if we just do

```ts
expect( screen.queryByText( /storage/i ) ).not.toBeInTheDocument() 
```

we can't be sure if the test pass because the banner never shows, or because somehow the component is still loading by the time the test is testing the assertion.

Comment (p-jackson) [reply client/dashboard/sites/overview/test/storage-warning-banner.tsx]: Yeah this is basically the one case where I care about double-checking that the component has finished loading. Testing the non-existence of some content is fraught enough. It could very easily end up writing an assertion that simply passes every time.

In fact, the test is so flaky, that if the component's copy changed then the test could end up auto-passing no matter what.

You know what, a better test would be to first test the banner is there, then test that the component is able to correctly re-render and remove the banner.

File: client/dashboard/app/router.tsx (hunk 9 above)
Comment (fushar) [reply client/dashboard/app/router.tsx]: Can we put the `await` at the end so that the Promises in the `preload` block here have chance to run?

File: client/dashboard/sites/overview/storage-warning-banner.tsx [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+import { userPreferenceQuery, userPreferenceMutation } from '../../app/queries/me-preferences';
+import { siteMediaStorageQuery } from '../../app/queries/site-media-storage';
+import Notice from '../../components/notice';
+import UpsellCTAButton from '../../components/upsell-cta-button';
+import { getStorageAlertLevel } from '../../utils/site-storage';
+import type { Site } from '../../data/types';
+
+export function StorageWarningBanner( { site }: { site: Site } ) {
+	const { data: mediaStorage } = useSuspenseQuery( siteMediaStorageQuery( site.ID ) );
+	const { data: isDismissedPersisted } = useSuspenseQuery(
```
Comment (fushar) [inline client/dashboard/sites/overview/storage-warning-banner.tsx:13]: Should we ensure the data in the routes, given that we're using `useSuspenseQuery()` here?

File: client/dashboard/utils/site-storage.ts (hunk 8 above)
Comment (fushar) [reply client/dashboard/utils/site-storage.ts]: Is this done, or are we missing some commits 🤔

File: client/dashboard/app/router.tsx (hunk 9 above)
Comment (p-jackson) [reply client/dashboard/app/router.tsx]: I've moved the `await` so that preloading starts before we block for the storage request ea497ed5f5be9859d59f366c233360109b1b52c5.

I'm still not sure whether this is better, since maybe we'd want there to be less network contention while we're running the blocking request 🤔 We've probably got multiple simultaneous connections going to `public-api` though, so probably not a problem.

File: client/dashboard/utils/site-storage.ts (hunk 8 above)
Comment (p-jackson) [reply client/dashboard/utils/site-storage.ts]: I don't know what happened to it! 😖 I guess I overwrote it locally? 🤷 
Done now ddc8072e21125a599fd5c74e4f9c962cd9e2626e

File: client/dashboard/sites/overview/storage-warning-banner.tsx (hunk 10 above)
Comment (p-jackson) [reply client/dashboard/sites/overview/storage-warning-banner.tsx:13]: Whoops I mis-read the router code. The preferences are already ensure'd in the *sites* route, not the *site* route 🤦 
Done in f335f330dc0

When I test by deleting the `REACT_QUERY_OFFLINE_CACHE` object I really am getting a clean loading experience of the overview page now.

File: client/dashboard/sites/overview/test/storage-warning-banner.tsx (hunk 3 above)
Comment (p-jackson) [reply client/dashboard/sites/overview/test/storage-warning-banner.tsx]: Ok, now the unit tests are _really_ good 😅 147c7f8ef0c

No longer needing to assert that the component isn't loading. Instead trying to focus on asserting things that users can actually see, like the banner being there and then not being there.

We also don't need a special test just to check that a dismissed banner doesn't appear. We already have the last test which runs through the whole user flow of clicking the dismiss button and the banner disappearing.

I had to rearrange the `test-utils` file a bit, but it was just to make the testing-library types happy.

File: client/dashboard/sites/overview/test/storage-warning-banner.tsx [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+		.reply( 200, { calypso_preferences: preferences } )
+		.post( '/rest/v1.1/me/preferences' )
+		.reply( 200, ( _uri, body ) => body );
+
+	return { mediaStorage, scope };
+}
+
+function renderTestBanner() {
+	// The only site field used by the banner is ID.
+	const testSite = { ID: siteId };
```
Comment (fushar) [inline client/dashboard/sites/overview/test/storage-warning-banner.tsx:49]: I usually just use `const testSite = { ID: siteId } as Site;` instead of `any` below; I think it should also work fine.

**Review by fushar — APPROVED** [review]
Tested; looking really good now. Thanks for working on the tests!

**Review by arthur791004 — APPROVED** [review]

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/19731202

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @p-jackson for including a screenshot in the description! This is really helpful for our translators.
